// amplify-provider-awscloudformation/lib/export-resources.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.run = void 0;
const amplify_cli_core_1 = require("amplify-cli-core");
const resource_export_1 = require("./resource-package/resource-export");
const path = __importStar(require("path"));
const amplify_prompts_1 = require("amplify-prompts");
const fs = __importStar(require("fs-extra"));
const ora_1 = __importDefault(require("ora"));
const backup = 'backup';
const lodash_1 = __importDefault(require("lodash"));
const rimraf_1 = __importDefault(require("rimraf"));
const amplify_cli_core_2 = require("amplify-cli-core");
async function run(context, resourceDefinition, exportPath) {
    const resolvedExportDir = amplify_cli_core_2.validateExportDirectoryPath(exportPath, amplify_cli_core_1.PathConstants.DefaultExportFolder);
    const { projectName } = amplify_cli_core_1.stateManager.getProjectConfig();
    const amplifyExportFolder = path.join(resolvedExportDir, `amplify-export-${projectName}`);
    const proceed = await checkForExistingExport(amplifyExportFolder);
    if (proceed) {
        await createBackup(amplifyExportFolder);
        deleteFolder(amplifyExportFolder);
    }
    else {
        return;
    }
    const spinner = ora_1.default('Exporting...');
    spinner.start();
    try {
        const resourceExport = new resource_export_1.ResourceExport(context, amplifyExportFolder);
        spinner.text = 'Building and packaging resources';
        const packagedResources = await resourceExport.packageBuildWriteResources(resourceDefinition);
        spinner.text = `Writing resources`;
        await resourceExport.writeResourcesToDestination(packagedResources);
        spinner.text = `Writing Cloudformation`;
        const { stackParameters, transformedResources } = await resourceExport.generateAndTransformCfnResources(packagedResources);
        spinner.text = `Generating and writing root stack`;
        const extractedParameters = await resourceExport.generateAndWriteRootStack(stackParameters);
        const parameters = resourceExport.fixNestedStackParameters(transformedResources, extractedParameters);
        spinner.text = `Generating export manifest`;
        writeExportManifest(parameters, resolvedExportDir, amplifyExportFolder);
        spinner.text = `Generating category stack mappings`;
        createCategoryStackMapping(transformedResources, amplifyExportFolder);
        spinner.text = 'Generating export tag file';
        createTagsFile(amplifyExportFolder);
        spinner.text = 'Setting permissions';
        await setPermissions(amplifyExportFolder);
        spinner.succeed();
        amplify_prompts_1.printer.blankLine();
        amplify_prompts_1.printer.success('Successfully exported');
        amplify_prompts_1.printer.info('Next steps:');
        amplify_prompts_1.printer.info('You can now integrate your Amplify Backend into your CDK App');
        amplify_prompts_1.printer.info('Install the "Amplify Exported Backend" CDK Construct by running "npm i @aws-amplify/cdk-exported-backend" in your CDK app');
        amplify_prompts_1.printer.info('For more information: docs.amplify.aws/cli/usage/export-to-cdk');
        amplify_prompts_1.printer.blankLine();
    }
    catch (ex) {
        revertToBackup(amplifyExportFolder);
        spinner.fail();
        throw ex;
    }
    finally {
        removeBackup(amplifyExportFolder);
        spinner.stop();
    }
}
exports.run = run;
async function setPermissions(amplifyExportFolder) {
    await fs.chmod(amplifyExportFolder, 0o700);
}
function createTagsFile(exportPath) {
    const hydratedTags = amplify_cli_core_1.stateManager.getHydratedTags(undefined, true);
    amplify_cli_core_1.JSONUtilities.writeJson(path.join(exportPath, amplify_cli_core_1.PathConstants.ExportTagsJsonFileName), hydratedTags.map(tag => ({
        key: tag.Key,
        value: tag.Value,
    })));
}
function createCategoryStackMapping(resources, amplifyExportFolder) {
    amplify_cli_core_1.JSONUtilities.writeJson(path.join(amplifyExportFolder, amplify_cli_core_1.PathConstants.ExportCategoryStackMappingJsonFilename), resources.map(r => {
        return lodash_1.default.pick(r, ['category', 'resourceName', 'service']);
    }));
}
async function checkForExistingExport(amplifyExportFolder) {
    let proceed = true;
    if (fs.existsSync(amplifyExportFolder)) {
        proceed = await amplify_prompts_1.prompter.yesOrNo(`Existing files at ${amplifyExportFolder} will be deleted and new files will be generated, continue?`, true);
    }
    await fs.ensureDir(amplifyExportFolder);
    return proceed;
}
function deleteFolder(directoryPath) {
    if (fs.existsSync(directoryPath)) {
        rimraf_1.default.sync(directoryPath);
    }
}
async function removeBackup(amplifyExportFolder) {
    if (fs.existsSync(`${amplifyExportFolder}-${backup}`)) {
        deleteFolder(`${amplifyExportFolder}-${backup}`);
    }
}
async function revertToBackup(amplifyExportFolder) {
    if (fs.existsSync(`${amplifyExportFolder}-${backup}`)) {
        await fs.copy(`${amplifyExportFolder}-${backup}`, amplifyExportFolder);
    }
}
async function createBackup(amplifyExportFolder) {
    await fs.copy(amplifyExportFolder, `${amplifyExportFolder}-${backup}`);
}
function writeExportManifest(stackParameters, exportPath, amplifyExportFolder) {
    const rootStackParametersKey = lodash_1.default.first(Object.keys(stackParameters));
    const manifestJson = {
        stackName: rootStackParametersKey,
        props: transformManifestParameters(stackParameters[rootStackParametersKey], amplifyExportFolder),
    };
    amplify_cli_core_1.JSONUtilities.writeJson(path.join(amplifyExportFolder, amplify_cli_core_1.PathConstants.ExportManifestJsonFilename), manifestJson);
}
function transformManifestParameters(stackParameters, exportPath) {
    if (stackParameters) {
        const manifest = {
            templateFile: path.relative(exportPath, stackParameters.destination),
            parameters: stackParameters.parameters,
            preserveLogicalIds: true,
            loadNestedStacks: {},
        };
        if (!stackParameters.nestedStacks) {
            return manifest;
        }
        Object.keys(stackParameters.nestedStacks)
            .sort()
            .forEach(key => {
            manifest['loadNestedStacks'][key] = transformManifestParameters(stackParameters.nestedStacks[key], exportPath);
        });
        return manifest;
    }
}
