// amplify-headless-interface/lib/interface/geo/add.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MapStyle = exports.AccessType = exports.PricingPlan = void 0;
var PricingPlan;
(function (PricingPlan) {
    PricingPlan["RequestBasedUsage"] = "RequestBasedUsage";
    PricingPlan["MobileAssetTracking"] = "MobileAssetTracking";
    PricingPlan["MobileAssetManagement"] = "MobileAssetManagement";
})(PricingPlan = exports.PricingPlan || (exports.PricingPlan = {}));
var AccessType;
(function (AccessType) {
    AccessType["AuthorizedUsers"] = "AuthorizedUsers";
    AccessType["AuthorizedAndGuestUsers"] = "AuthorizedAndGuestUsers";
})(AccessType = exports.AccessType || (exports.AccessType = {}));
var MapStyle;
(function (MapStyle) {
    MapStyle["VectorEsriNavigation"] = "VectorEsriNavigation";
    MapStyle["VectorEsriStreets"] = "VectorEsriStreets";
    MapStyle["VectorEsriTopographic"] = "VectorEsriTopographic";
    MapStyle["VectorEsriDarkGrayCanvas"] = "VectorEsriDarkGrayCanvas";
    MapStyle["VectorEsriLightGrayCanvas"] = "VectorEsriLightGrayCanvas";
    MapStyle["VectorHereBerlin"] = "VectorHereBerlin";
})(MapStyle = exports.MapStyle || (exports.MapStyle = {}));
