// amplify-provider-awscloudformation/lib/root-stack-builder/root-stack-transform.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmplifyRootStackTransform = void 0;
const cdk = __importStar(require("@aws-cdk/core"));
const core_1 = require("@aws-cdk/core");
const amplify_cli_core_1 = require("amplify-cli-core");
const amplifyPrinter = __importStar(require("amplify-prompts"));
const amplify_prompts_1 = require("amplify-prompts");
const fs = __importStar(require("fs-extra"));
const os_1 = __importDefault(require("os"));
const path = __importStar(require("path"));
const vm = __importStar(require("vm2"));
const root_stack_builder_1 = require("./root-stack-builder");
const stack_synthesizer_1 = require("./stack-synthesizer");
class AmplifyRootStackTransform {
    constructor(resourceName) {
        this.applyOverride = async () => {
            const backendDir = amplify_cli_core_1.pathManager.getBackendDirPath();
            const overrideFilePath = path.join(backendDir, this._resourceName);
            let isBuild = false;
            try {
                isBuild = await amplify_cli_core_1.buildOverrideDir(backendDir, overrideFilePath);
            }
            catch (error) {
                amplifyPrinter.printer.debug(`Skipping build as ${error.message}`);
            }
            if (isBuild) {
                const overrideCode = await fs.readFile(path.join(overrideFilePath, 'build', 'override.js'), 'utf-8').catch(() => {
                    amplifyPrinter.formatter.list(['No override File Found', `To override ${this._resourceName} run amplify override auth`]);
                    return '';
                });
                const sandboxNode = new vm.NodeVM({
                    console: 'inherit',
                    timeout: 5000,
                    sandbox: {},
                    require: {
                        context: 'sandbox',
                        builtin: ['path'],
                        external: true,
                    },
                });
                try {
                    sandboxNode.run(overrideCode).override(this._rootTemplateObj);
                }
                catch (err) {
                    const error = new Error(`Skipping override due to ${err}${os_1.default.EOL}`);
                    amplify_prompts_1.printer.error(`${error}`);
                    error.stack = undefined;
                    throw error;
                }
            }
        };
        this.generateRootStackTemplate = async () => {
            this._rootTemplateObj = new root_stack_builder_1.AmplifyRootStack(this.app, 'AmplifyRootStack', { synthesizer: this._synthesizer });
            this._rootTemplateObj.addCfnParameter({
                type: 'String',
                description: 'Name of the common deployment bucket provided by the parent stack',
                default: 'DeploymentBucket',
            }, 'DeploymentBucketName');
            this._rootTemplateObj.addCfnParameter({
                type: 'String',
                description: 'Name of the common deployment bucket provided by the parent stack',
                default: 'AuthRoleName',
            }, 'AuthRoleName');
            this._rootTemplateObj.addCfnParameter({
                type: 'String',
                description: 'Name of the common deployment bucket provided by the parent stack',
                default: 'UnAuthRoleName',
            }, 'UnauthRoleName');
            this._rootTemplateObj.addCfnOutput({
                description: 'CloudFormation provider root stack Region',
                value: cdk.Fn.ref('AWS::Region'),
                exportName: cdk.Fn.sub('${AWS::StackName}-Region'),
            }, 'Region');
            this._rootTemplateObj.addCfnOutput({
                description: 'CloudFormation provider root stack ID',
                value: cdk.Fn.ref('AWS::StackName'),
                exportName: cdk.Fn.sub('${AWS::StackName}-StackName'),
            }, 'StackName');
            this._rootTemplateObj.addCfnOutput({
                description: 'CloudFormation provider root stack name',
                value: cdk.Fn.ref('AWS::StackId'),
                exportName: cdk.Fn.sub('${AWS::StackName}-StackId'),
            }, 'StackId');
            this._rootTemplateObj.addCfnOutput({
                value: cdk.Fn.getAtt('AuthRole', 'Arn').toString(),
            }, 'AuthRoleArn');
            this._rootTemplateObj.addCfnOutput({
                value: cdk.Fn.getAtt('UnauthRole', 'Arn').toString(),
            }, 'UnauthRoleArn');
            await this._rootTemplateObj.generateRootStackResources();
            this._rootTemplateObjOutputs = new root_stack_builder_1.AmplifyRootStackOutputs(this.app, 'AmplifyRootStackOutputs', {
                synthesizer: this._synthesizerOutputs,
            });
            this._rootTemplateObjOutputs.addCfnOutput({
                description: 'CloudFormation provider root stack deployment bucket name',
                value: cdk.Fn.ref('DeploymentBucketName'),
                exportName: cdk.Fn.sub('${AWS::StackName}-DeploymentBucketName'),
            }, 'DeploymentBucketName');
            this._rootTemplateObjOutputs.addCfnOutput({
                value: cdk.Fn.ref('AuthRole'),
            }, 'AuthRoleName');
            this._rootTemplateObjOutputs.addCfnOutput({
                value: cdk.Fn.ref('UnauthRole'),
            }, 'UnauthRoleName');
        };
        this.synthesizeTemplates = async () => {
            var _a;
            (_a = this.app) === null || _a === void 0 ? void 0 : _a.synth();
            const templates = this._synthesizer.collectStacks();
            const templatesOutput = this._synthesizerOutputs.collectStacks();
            const cfnRootStack = templates.get('AmplifyRootStack');
            const cfnRootStackOutputs = templatesOutput.get('AmplifyRootStackOutputs');
            Object.assign(cfnRootStack.Outputs, cfnRootStackOutputs.Outputs);
            return cfnRootStack;
        };
        this.saveBuildFiles = async (context, template) => {
            const rootStackFileName = `root-cloudformation-stack.json`;
            const rootstackFilePath = path.join(amplify_cli_core_1.pathManager.getBackendDirPath(), this._resourceName, 'build', rootStackFileName);
            await amplify_cli_core_1.writeCFNTemplate(template, rootstackFilePath, {
                templateFormat: amplify_cli_core_1.CFNTemplateFormat.JSON,
            });
        };
        this._resourceName = resourceName;
        this._synthesizer = new stack_synthesizer_1.RootStackSythesizer();
        this.app = new core_1.App();
        this._synthesizerOutputs = new stack_synthesizer_1.RootStackSythesizer();
    }
    async transform(context) {
        await this.generateRootStackTemplate();
        if (context.input.command !== 'init') {
            await this.applyOverride();
        }
        const template = await this.synthesizeTemplates();
        if (context.input.command !== 'init') {
            await this.saveBuildFiles(context, template);
        }
        return template;
    }
    getRootStack() {
        if (this._rootTemplateObj != null) {
            return this._rootTemplateObj;
        }
        else {
            throw new Error('Root Stack Template doesnt exist');
        }
    }
}
exports.AmplifyRootStackTransform = AmplifyRootStackTransform;
