// amplify-provider-awscloudformation/lib/root-stack-builder/root-stack-builder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmplifyRootStackOutputs = exports.AmplifyRootStack = void 0;
const cdk = __importStar(require("@aws-cdk/core"));
const s3 = __importStar(require("@aws-cdk/aws-s3"));
const iam = __importStar(require("@aws-cdk/aws-iam"));
const CFN_TEMPLATE_FORMAT_VERSION = '2010-09-09';
const ROOT_CFN_DESCRIPTION = 'Root Stack for AWS Amplify CLI';
class AmplifyRootStack extends cdk.Stack {
    constructor(scope, id, props) {
        super(scope, id, props);
        this._cfnParameterMap = new Map();
        this.generateRootStackResources = async () => {
            this.deploymentBucket = new s3.CfnBucket(this, 'DeploymentBucket', {
                bucketName: this._cfnParameterMap.get('DeploymentBucketName').valueAsString,
            });
            this.deploymentBucket.applyRemovalPolicy(cdk.RemovalPolicy.RETAIN);
            this.authRole = new iam.CfnRole(this, 'AuthRole', {
                roleName: this._cfnParameterMap.get('AuthRoleName').valueAsString,
                assumeRolePolicyDocument: {
                    Version: '2012-10-17',
                    Statement: [
                        {
                            Sid: '',
                            Effect: 'Deny',
                            Principal: {
                                Federated: 'cognito-identity.amazonaws.com',
                            },
                            Action: 'sts:AssumeRoleWithWebIdentity',
                        },
                    ],
                },
            });
            this.unauthRole = new iam.CfnRole(this, 'UnauthRole', {
                roleName: this._cfnParameterMap.get('UnauthRoleName').valueAsString,
                assumeRolePolicyDocument: {
                    Version: '2012-10-17',
                    Statement: [
                        {
                            Sid: '',
                            Effect: 'Deny',
                            Principal: {
                                Federated: 'cognito-identity.amazonaws.com',
                            },
                            Action: 'sts:AssumeRoleWithWebIdentity',
                        },
                    ],
                },
            });
        };
        this.renderCloudFormationTemplate = (_) => {
            return JSON.stringify(this._toCloudFormation(), undefined, 2);
        };
        this._scope = scope;
        this.templateOptions.templateFormatVersion = CFN_TEMPLATE_FORMAT_VERSION;
        this.templateOptions.description = ROOT_CFN_DESCRIPTION;
    }
    addCfnOutput(props, logicalId) {
        try {
            new cdk.CfnOutput(this, logicalId, props);
        }
        catch (error) {
            throw new Error(error);
        }
    }
    addCfnMapping(props, logicalId) {
        try {
            new cdk.CfnMapping(this, logicalId, props);
        }
        catch (error) {
            throw new Error(error);
        }
    }
    addCfnCondition(props, logicalId) {
        try {
            new cdk.CfnCondition(this, logicalId, props);
        }
        catch (error) {
            throw new Error(error);
        }
    }
    addCfnResource(props, logicalId) {
        try {
            new cdk.CfnResource(this, logicalId, props);
        }
        catch (error) {
            throw new Error(error);
        }
    }
    addCfnParameter(props, logicalId) {
        try {
            if (this._cfnParameterMap.has(logicalId)) {
                throw new Error('logical Id already Exists');
            }
            this._cfnParameterMap.set(logicalId, new cdk.CfnParameter(this, logicalId, props));
        }
        catch (error) {
            throw new Error(error);
        }
    }
    getCfnParameter(logicalId) {
        if (this._cfnParameterMap.has(logicalId)) {
            return this._cfnParameterMap.get(logicalId);
        }
        else {
            throw new Error(`Cfn Parameter with LogicalId ${logicalId} doesnt exist`);
        }
    }
}
exports.AmplifyRootStack = AmplifyRootStack;
class AmplifyRootStackOutputs extends cdk.Stack {
    constructor(scope, id, props) {
        super(scope, id, props);
        this.renderCloudFormationTemplate = (_) => {
            return JSON.stringify(this._toCloudFormation(), undefined, 2);
        };
    }
    addCfnParameter(props, logicalId) {
        throw new Error('Method not implemented.');
    }
    addCfnOutput(props, logicalId) {
        try {
            new cdk.CfnOutput(this, logicalId, props);
        }
        catch (error) {
            throw new Error(error);
        }
    }
    addCfnMapping(props, logicalId) {
        throw new Error('Method not implemented.');
    }
    addCfnCondition(props, logicalId) {
        throw new Error('Method not implemented.');
    }
    addCfnResource(props, logicalId) {
        throw new Error('Method not implemented.');
    }
}
exports.AmplifyRootStackOutputs = AmplifyRootStackOutputs;
