// amplify-provider-awscloudformation/lib/root-stack-builder/stack-synthesizer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RootStackSythesizer = void 0;
const core_1 = require("@aws-cdk/core");
const root_stack_builder_1 = require("./root-stack-builder");
class RootStackSythesizer extends core_1.LegacyStackSynthesizer {
    constructor() {
        super(...arguments);
        this.stacks = new Map();
        this.getStack = (stackName) => {
            if (this.stacks.has(stackName)) {
                return this.stacks.get(stackName);
            }
            throw new Error(`Stack ${stackName} is not created`);
        };
    }
    synthesizeStackTemplate(stack, session) {
        if (stack instanceof root_stack_builder_1.AmplifyRootStack || stack instanceof root_stack_builder_1.AmplifyRootStackOutputs) {
            this.addStack(stack);
            const template = stack.renderCloudFormationTemplate(session);
            const templateName = stack.node.id;
            this.setStackAsset(templateName, template);
        }
        else {
            throw new Error('Error synthesizing the template. Expected Stack to be either instance of AmplifyRootStack');
        }
    }
    setStackAsset(templateName, template) {
        RootStackSythesizer.stackAssets.set(templateName, JSON.parse(template));
    }
    collectStacks() {
        return new Map(RootStackSythesizer.stackAssets.entries());
    }
    addStack(stack) {
        this.stacks.set(stack.node.id, stack);
    }
}
exports.RootStackSythesizer = RootStackSythesizer;
RootStackSythesizer.stackAssets = new Map();
