// @aws-amplify/graphql-model-transformer/lib/resolvers/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateAuthExpressionForSandboxMode = exports.generateResolverKey = exports.generateDefaultResponseMappingTemplate = exports.generateConditionSlot = void 0;
const graphql_mapping_template_1 = require("graphql-mapping-template");
const definitions_1 = require("../definitions");
const API_KEY = 'API Key Authorization';
const generateConditionSlot = (inputConditionObjectName, conditionOutputVariableName) => {
    const statements = [
        graphql_mapping_template_1.comment('Start condition block'),
        graphql_mapping_template_1.iff(graphql_mapping_template_1.and([graphql_mapping_template_1.ref(inputConditionObjectName), graphql_mapping_template_1.notEquals(graphql_mapping_template_1.methodCall(graphql_mapping_template_1.ref(`${inputConditionObjectName}.size`)), graphql_mapping_template_1.int(0))]), graphql_mapping_template_1.compoundExpression([
            graphql_mapping_template_1.set(graphql_mapping_template_1.ref('mergedConditions'), graphql_mapping_template_1.obj({ and: graphql_mapping_template_1.ref(inputConditionObjectName) })),
            graphql_mapping_template_1.set(graphql_mapping_template_1.ref(conditionOutputVariableName), graphql_mapping_template_1.methodCall(graphql_mapping_template_1.ref('util.parseJson'), graphql_mapping_template_1.methodCall(graphql_mapping_template_1.ref('util.transform.toDynamoDBConditionExpression'), graphql_mapping_template_1.ref('mergedConditions')))),
            graphql_mapping_template_1.iff(graphql_mapping_template_1.and([
                graphql_mapping_template_1.ref(`${conditionOutputVariableName}.expressionValues`),
                graphql_mapping_template_1.equals(graphql_mapping_template_1.methodCall(graphql_mapping_template_1.ref(`${conditionOutputVariableName}.expressionValues.size`)), graphql_mapping_template_1.int(0)),
            ]), graphql_mapping_template_1.set(graphql_mapping_template_1.ref(conditionOutputVariableName), graphql_mapping_template_1.obj({
                expression: graphql_mapping_template_1.ref(`${conditionOutputVariableName}.expression`),
                expressionNames: graphql_mapping_template_1.ref(`${conditionOutputVariableName}.expressionNames`),
            }))),
            graphql_mapping_template_1.comment('End condition block'),
        ])),
    ];
    return graphql_mapping_template_1.compoundExpression(statements);
};
exports.generateConditionSlot = generateConditionSlot;
const generateDefaultResponseMappingTemplate = (isSyncEnabled, mutation = false) => {
    const statements = [];
    if (mutation)
        statements.push(graphql_mapping_template_1.qref(graphql_mapping_template_1.methodCall(graphql_mapping_template_1.ref('ctx.result.put'), graphql_mapping_template_1.str(definitions_1.OPERATION_KEY), graphql_mapping_template_1.str('Mutation'))));
    if (isSyncEnabled) {
        statements.push(graphql_mapping_template_1.ifElse(graphql_mapping_template_1.ref('ctx.error'), graphql_mapping_template_1.methodCall(graphql_mapping_template_1.ref('util.error'), graphql_mapping_template_1.ref('ctx.error.message'), graphql_mapping_template_1.ref('ctx.error.type'), graphql_mapping_template_1.ref('ctx.result')), graphql_mapping_template_1.toJson(graphql_mapping_template_1.ref('ctx.result'))));
    }
    else {
        statements.push(graphql_mapping_template_1.ifElse(graphql_mapping_template_1.ref('ctx.error'), graphql_mapping_template_1.methodCall(graphql_mapping_template_1.ref('util.error'), graphql_mapping_template_1.ref('ctx.error.message'), graphql_mapping_template_1.ref('ctx.error.type')), graphql_mapping_template_1.toJson(graphql_mapping_template_1.ref('ctx.result'))));
    }
    return graphql_mapping_template_1.printBlock('ResponseTemplate')(graphql_mapping_template_1.compoundExpression(statements));
};
exports.generateDefaultResponseMappingTemplate = generateDefaultResponseMappingTemplate;
const generateResolverKey = (typeName, fieldName) => {
    return `${typeName}.${fieldName}`;
};
exports.generateResolverKey = generateResolverKey;
const generateAuthExpressionForSandboxMode = (enabled) => {
    let exp;
    if (enabled)
        exp = graphql_mapping_template_1.iff(graphql_mapping_template_1.notEquals(graphql_mapping_template_1.methodCall(graphql_mapping_template_1.ref('util.authType')), graphql_mapping_template_1.str(API_KEY)), graphql_mapping_template_1.methodCall(graphql_mapping_template_1.ref('util.unauthorized')));
    else
        exp = graphql_mapping_template_1.methodCall(graphql_mapping_template_1.ref('util.unauthorized'));
    return graphql_mapping_template_1.printBlock(`Sandbox Mode ${enabled ? 'Enabled' : 'Disabled'}`)(graphql_mapping_template_1.compoundExpression([graphql_mapping_template_1.iff(graphql_mapping_template_1.not(graphql_mapping_template_1.ref('ctx.stash.get("hasAuth")')), exp), graphql_mapping_template_1.toJson(graphql_mapping_template_1.obj({}))]));
};
exports.generateAuthExpressionForSandboxMode = generateAuthExpressionForSandboxMode;
