// @aws-amplify/graphql-transformer-migrator/lib/migrators/generators/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAuthRule = exports.createListValueNode = exports.createArgumentNode = exports.createDirectiveNode = exports.createNameNode = void 0;
const graphql_1 = require("graphql");
const auth_1 = require("../auth");
function createNameNode(name) {
    return {
        kind: 'Name',
        value: name,
    };
}
exports.createNameNode = createNameNode;
function createDirectiveNode(name, args) {
    return {
        kind: 'Directive',
        name: createNameNode(name),
        arguments: args,
    };
}
exports.createDirectiveNode = createDirectiveNode;
function createArgumentNode(name, value) {
    return {
        kind: 'Argument',
        name: createNameNode(name),
        value: value,
    };
}
exports.createArgumentNode = createArgumentNode;
function createListValueNode(values) {
    return {
        kind: 'ListValue',
        values: values,
    };
}
exports.createListValueNode = createListValueNode;
function createAuthRule(strategy, provider, operations) {
    let rule = `{allow: ${strategy}`;
    if (provider && provider !== auth_1.defaultProviderMap.get(strategy)) {
        rule += `, provider: ${provider}`;
    }
    if (operations && operations.length !== 4) {
        rule += `, operations: [${operations.join(', ')}]`;
    }
    rule += '}';
    return graphql_1.parseValue(rule);
}
exports.createAuthRule = createAuthRule;
