// @aws-amplify/amplify-category-storage/lib/provider-utils/awscloudformation/cdk-stack-builder/types.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmplifyBuildParamsPermissions = exports.AmplifyResourceCfnStack = void 0;
const cdk = __importStar(require("@aws-cdk/core"));
class AmplifyResourceCfnStack extends cdk.Stack {
    constructor(scope, id) {
        super(scope, id, undefined);
        this._cfnParameterMap = new Map();
        this.renderCloudFormationTemplate = () => {
            return this._toCloudFormation();
        };
    }
    addCfnOutput(props, logicalId) {
        try {
            new cdk.CfnOutput(this, logicalId, props);
        }
        catch (error) {
            throw error;
        }
    }
    addCfnMapping(props, logicalId) {
        try {
            new cdk.CfnMapping(this, logicalId, props);
        }
        catch (error) {
            throw error;
        }
    }
    addCfnCondition(props, logicalId) {
        try {
            new cdk.CfnCondition(this, logicalId, props);
        }
        catch (error) {
            throw error;
        }
    }
    addCfnResource(props, logicalId) {
        try {
            return new cdk.CfnResource(this, logicalId, props);
        }
        catch (error) {
            throw error;
        }
    }
    addCfnParameter(props, logicalId) {
        try {
            if (this._cfnParameterMap.has(logicalId)) {
                throw new Error('logical Id already Exists');
            }
            this._cfnParameterMap.set(logicalId, new cdk.CfnParameter(this, logicalId, props));
        }
        catch (error) {
            throw error;
        }
    }
}
exports.AmplifyResourceCfnStack = AmplifyResourceCfnStack;
var AmplifyBuildParamsPermissions;
(function (AmplifyBuildParamsPermissions) {
    AmplifyBuildParamsPermissions["ALLOW"] = "ALLOW";
    AmplifyBuildParamsPermissions["DISALLOW"] = "DISALLOW";
})(AmplifyBuildParamsPermissions = exports.AmplifyBuildParamsPermissions || (exports.AmplifyBuildParamsPermissions = {}));
