// @aws-amplify/amplify-category-auth/lib/provider-utils/awscloudformation/auth-stack-builder/auth-stack-transform.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmplifyAuthTransform = void 0;
const amplify_cli_core_1 = require("amplify-cli-core");
const auth_cognito_stack_builder_1 = require("./auth-cognito-stack-builder");
const stack_synthesizer_1 = require("./stack-synthesizer");
const cdk = __importStar(require("@aws-cdk/core"));
const auth_input_state_1 = require("../auth-inputs-manager/auth-input-state");
const lodash_1 = __importDefault(require("lodash"));
const path = __importStar(require("path"));
const amplify_prompts_1 = require("amplify-prompts");
const generate_auth_trigger_template_1 = require("../utils/generate-auth-trigger-template");
const synthesize_resources_1 = require("../utils/synthesize-resources");
const awsCognito_user_input_types_1 = require("../service-walkthrough-types/awsCognito-user-input-types");
const vm = __importStar(require("vm2"));
const fs = __importStar(require("fs-extra"));
const os_1 = __importDefault(require("os"));
class AmplifyAuthTransform extends amplify_cli_core_1.AmplifyCategoryTransform {
    constructor(resourceName) {
        super(resourceName);
        this.applyOverride = async () => {
            const backendDir = amplify_cli_core_1.pathManager.getBackendDirPath();
            const overrideDir = path.join(backendDir, this._category, this.resourceName);
            const isBuild = await amplify_cli_core_1.buildOverrideDir(backendDir, overrideDir).catch(error => {
                amplify_prompts_1.printer.debug(`Skipping build as ${error.message}`);
                return false;
            });
            if (isBuild) {
                const overrideCode = await fs.readFile(path.join(overrideDir, 'build', 'override.js'), 'utf-8').catch(() => {
                    amplify_prompts_1.formatter.list(['No override File Found', `To override ${this.resourceName} run amplify override auth`]);
                    return '';
                });
                const sandboxNode = new vm.NodeVM({
                    console: 'inherit',
                    timeout: 5000,
                    sandbox: {},
                    require: {
                        context: 'sandbox',
                        builtin: ['path'],
                        external: true,
                    },
                });
                try {
                    await sandboxNode
                        .run(overrideCode, path.join(overrideDir, 'build', 'override.js'))
                        .override(this._authTemplateObj);
                }
                catch (err) {
                    const error = new Error(`Skipping override due to ${err}${os_1.default.EOL}`);
                    amplify_prompts_1.printer.error(`${error}`);
                    error.stack = undefined;
                    throw error;
                }
            }
        };
        this.generateStackProps = async (context) => {
            const roles = {
                authRoleArn: {
                    'Fn::GetAtt': ['AuthRole', 'Arn'],
                },
                unauthRoleArn: {
                    'Fn::GetAtt': ['UnauthRole', 'Arn'],
                },
            };
            let cognitoStackProps = {
                ...this._cliInputs.cognitoConfig,
                ...roles,
                breakCircularDependency: amplify_cli_core_1.FeatureFlags.getBoolean('auth.breakcirculardependency'),
                useEnabledMfas: amplify_cli_core_1.FeatureFlags.getBoolean('auth.useenabledmfas'),
                dependsOn: [],
            };
            const teamProviderobj = context.amplify.loadEnvResourceParameters(context, this._category, this.resourceName);
            if (!lodash_1.default.isEmpty(teamProviderobj)) {
                cognitoStackProps = Object.assign(cognitoStackProps, teamProviderobj);
            }
            if (!lodash_1.default.isEmpty(this._cliInputs.cognitoConfig.triggers)) {
                const permissions = await context.amplify.getTriggerPermissions(context, this._cliInputs.cognitoConfig.triggers, amplify_cli_core_1.AmplifyCategories.AUTH, this._cliInputs.cognitoConfig.resourceName);
                const triggerPermissions = permissions === null || permissions === void 0 ? void 0 : permissions.map((i) => JSON.parse(i));
                const dependsOnKeys = Object.keys(this._cliInputs.cognitoConfig.triggers).map(i => `${this._cliInputs.cognitoConfig.resourceName}${i}`);
                const dependsOn = context.amplify.dependsOnBlock(context, dependsOnKeys, 'Cognito');
                const keys = Object.keys(this._cliInputs.cognitoConfig.triggers);
                const authTriggerConnections = [];
                keys.forEach(key => {
                    let config = {
                        triggerType: key === 'PreSignup' ? 'PreSignUp' : key,
                        lambdaFunctionName: `${this.resourceName}${key}`,
                    };
                    authTriggerConnections.push(config);
                });
                cognitoStackProps = Object.assign(cognitoStackProps, { permissions: triggerPermissions, dependsOn, authTriggerConnections });
            }
            return cognitoStackProps;
        };
        this.synthesizeTemplates = async () => {
            this._app.synth();
            const templates = this._synthesizer.collectStacks();
            return templates.get('AmplifyAuthCongitoStack');
        };
        this.saveBuildFiles = async (context, template) => {
            const cognitoStackFileName = `${this.resourceName}-cloudformation-template.json`;
            const cognitostackFilePath = path.join(amplify_cli_core_1.pathManager.getBackendDirPath(), this._category, this.resourceName, 'build', cognitoStackFileName);
            await amplify_cli_core_1.writeCFNTemplate(template, cognitostackFilePath, {
                templateFormat: amplify_cli_core_1.CFNTemplateFormat.JSON,
            });
            await this.writeBuildFiles(context);
        };
        this.writeBuildFiles = async (context) => {
            const parametersJSONFilePath = path.join(amplify_cli_core_1.pathManager.getBackendDirPath(), this._category, this.resourceName, 'build', 'parameters.json');
            const roles = {
                authRoleArn: {
                    'Fn::GetAtt': ['AuthRole', 'Arn'],
                },
                unauthRoleArn: {
                    'Fn::GetAtt': ['UnauthRole', 'Arn'],
                },
            };
            let parameters = {
                ...this._cliInputs.cognitoConfig,
                ...roles,
                breakCircularDependency: this._cognitoStackProps.breakCircularDependency,
                useEnabledMfas: this._cognitoStackProps.useEnabledMfas,
                dependsOn: [],
            };
            if (this._cognitoStackProps.triggers && !lodash_1.default.isEmpty(this._cognitoStackProps.triggers)) {
                this._cognitoStackProps.triggers = JSON.stringify(this._cognitoStackProps.triggers);
                const triggerPermissions = this._cognitoStackProps.permissions.map(i => JSON.stringify(i));
                const dependsOn = this._cognitoStackProps.dependsOn;
                const authTriggerConnections = this._cognitoStackProps.authTriggerConnections.map(obj => {
                    const modifiedObj = lodash_1.default.omit(obj, ['lambdaFunctionArn']);
                    return JSON.stringify(modifiedObj);
                });
                parameters = Object.assign(parameters, {
                    permissions: triggerPermissions,
                    triggers: this._cognitoStackProps.triggers,
                    dependsOn,
                    authTriggerConnections,
                });
            }
            else if (lodash_1.default.isEmpty(this._cognitoStackProps.triggers)) {
                parameters = Object.assign(parameters, { triggers: JSON.stringify(this._cognitoStackProps.triggers) });
            }
            amplify_cli_core_1.JSONUtilities.writeJson(parametersJSONFilePath, parameters);
        };
        this.generateCfnOutputs = (props) => {
            const configureSMS = (props.autoVerifiedAttributes && props.autoVerifiedAttributes.includes('phone_number')) ||
                (props.mfaConfiguration != 'OFF' && props.mfaTypes && props.mfaTypes.includes('SMS Text Message')) ||
                (props.requiredAttributes && props.requiredAttributes.includes('phone_number')) ||
                (props.usernameAttributes && props.usernameAttributes.includes(awsCognito_user_input_types_1.AttributeType.PHONE_NUMBER));
            if (props.authSelections === 'identityPoolAndUserPool' || props.authSelections == 'identityPoolOnly') {
                this._authTemplateObj.addCfnOutput({
                    value: cdk.Fn.ref('IdentityPool'),
                    description: 'Id for the identity pool',
                }, 'IdentityPoolId');
                this._authTemplateObj.addCfnOutput({
                    value: cdk.Fn.getAtt('IdentityPool', 'Name').toString(),
                }, 'IdentityPoolName');
            }
            if (props.hostedUIDomainName) {
                this._authTemplateObj.addCfnOutput({
                    value: cdk.Fn.conditionIf('ShouldNotCreateEnvResources', cdk.Fn.ref('hostedUIDomainName'), cdk.Fn.join('-', [cdk.Fn.ref('hostedUIDomainName'), cdk.Fn.ref('env')])).toString(),
                }, 'HostedUIDomain');
            }
            if (props.oAuthMetadata) {
                this._authTemplateObj.addCfnOutput({
                    value: cdk.Fn.ref('oAuthMetadata'),
                }, 'OAuthMetadata');
            }
            if (props.authSelections !== 'identityPoolOnly') {
                this._authTemplateObj.addCfnOutput({
                    value: cdk.Fn.ref('UserPool'),
                    description: 'Id for the user pool',
                }, 'UserPoolId');
                this._authTemplateObj.addCfnOutput({
                    value: cdk.Fn.getAtt('UserPool', 'Arn').toString(),
                    description: 'Arn for the user pool',
                }, 'UserPoolArn');
                this._authTemplateObj.addCfnOutput({
                    value: cdk.Fn.ref('userPoolName'),
                }, 'UserPoolName');
                this._authTemplateObj.addCfnOutput({
                    value: cdk.Fn.ref('UserPoolClientWeb'),
                    description: 'The user pool app client id for web',
                }, 'AppClientIDWeb');
                this._authTemplateObj.addCfnOutput({
                    value: cdk.Fn.ref('UserPoolClient'),
                    description: 'The user pool app client id',
                }, 'AppClientID');
                this._authTemplateObj.addCfnOutput({
                    value: cdk.Fn.getAtt('UserPoolClientInputs', 'appSecret').toString(),
                    condition: this._authTemplateObj.getCfnCondition('ShouldOutputAppClientSecrets'),
                }, 'AppClientSecret');
                if (!props.useEnabledMfas || configureSMS) {
                    this._authTemplateObj.addCfnOutput({
                        value: cdk.Fn.getAtt('SNSRole', 'Arn').toString(),
                        description: 'role arn',
                    }, 'CreatedSNSRole');
                }
                if (props.googleClientId) {
                    this._authTemplateObj.addCfnOutput({
                        value: cdk.Fn.ref('googleClientId'),
                    }, 'GoogleWebClient');
                }
                if (props.googleIos) {
                    this._authTemplateObj.addCfnOutput({
                        value: cdk.Fn.ref('googleIos'),
                    }, 'GoogleIOSClient');
                }
                if (props.googleAndroid) {
                    this._authTemplateObj.addCfnOutput({
                        value: cdk.Fn.ref('googleAndroid'),
                    }, 'GoogleAndroidClient');
                }
                if (props.facebookAppId) {
                    this._authTemplateObj.addCfnOutput({
                        value: cdk.Fn.ref('facebookAppId'),
                    }, 'FacebookWebClient');
                }
                if (props.amazonAppId) {
                    this._authTemplateObj.addCfnOutput({
                        value: cdk.Fn.ref('amazonAppId'),
                    }, 'AmazonWebClient');
                }
                if (props.appleAppId) {
                    this._authTemplateObj.addCfnOutput({
                        value: cdk.Fn.ref('appleAppId'),
                    }, 'AppleWebClient');
                }
            }
        };
        this.addCfnParameters = (props) => {
            this._authTemplateObj.addCfnParameter({
                type: 'String',
            }, 'env');
            if (!lodash_1.default.isEmpty(props.dependsOn)) {
                const dependsOn = props.dependsOn;
                dependsOn === null || dependsOn === void 0 ? void 0 : dependsOn.forEach(param => {
                    param.attributes.forEach(attribute => {
                        this._authTemplateObj.addCfnParameter({
                            type: 'String',
                            default: `${param.category}${param.resourceName}${attribute}`,
                        }, `${param.category}${param.resourceName}${attribute}`);
                    });
                });
            }
            for (const [key, value] of Object.entries(props)) {
                if (typeof value === 'string' || (typeof value === 'object' && !Array.isArray(value))) {
                    this._authTemplateObj.addCfnParameter({
                        type: 'String',
                    }, `${key}`);
                }
                if (typeof value === 'boolean') {
                    this._authTemplateObj.addCfnParameter({
                        type: 'String',
                    }, `${key}`);
                }
                if (typeof value === 'number') {
                    this._authTemplateObj.addCfnParameter({
                        type: 'String',
                    }, `${key}`);
                }
                if (value === 'parentStack') {
                    this._authTemplateObj.addCfnParameter({
                        type: 'String',
                    }, `${key}`);
                }
                if (Array.isArray(value)) {
                    this._authTemplateObj.addCfnParameter({
                        type: 'CommaDelimitedList',
                    }, `${key}`);
                }
            }
            if (Object.keys(props).includes('hostedUIProviderMeta') && !Object.keys(props).includes('hostedUIProviderCreds')) {
                this._authTemplateObj.addCfnParameter({
                    type: 'String',
                    default: '[]',
                }, 'hostedUIProviderCreds');
            }
        };
        this.addCfnConditions = (props) => {
            this._authTemplateObj.addCfnCondition({
                expression: cdk.Fn.conditionEquals(cdk.Fn.ref('env'), 'NONE'),
            }, 'ShouldNotCreateEnvResources');
            if (props.authSelections !== 'identityPoolOnly') {
                this._authTemplateObj.addCfnCondition({
                    expression: cdk.Fn.conditionEquals(cdk.Fn.ref('userpoolClientGenerateSecret'), true),
                }, 'ShouldOutputAppClientSecrets');
            }
        };
        this._synthesizer = new stack_synthesizer_1.AuthStackSythesizer();
        this._app = new cdk.App();
        this._category = amplify_cli_core_1.AmplifyCategories.AUTH;
        this._service = amplify_cli_core_1.AmplifySupportedService.COGNITO;
        this._authTemplateObj = new auth_cognito_stack_builder_1.AmplifyAuthCognitoStack(this._app, 'AmplifyAuthCongitoStack', { synthesizer: this._synthesizer });
    }
    async transform(context) {
        var _a;
        const cliState = new auth_input_state_1.AuthInputState(this.resourceName);
        this._cliInputs = cliState.getCLIInputPayload();
        this._cognitoStackProps = await this.generateStackProps(context);
        const resources = amplify_cli_core_1.stateManager.getMeta();
        if ((_a = resources.auth) === null || _a === void 0 ? void 0 : _a.userPoolGroups) {
            await synthesize_resources_1.updateUserPoolGroups(context, this._cognitoStackProps.resourceName, this._cognitoStackProps.userPoolGroupList);
        }
        else {
            await synthesize_resources_1.createUserPoolGroups(context, this._cognitoStackProps.resourceName, this._cognitoStackProps.userPoolGroupList);
        }
        if (this._cognitoStackProps.breakCircularDependency) {
            await generate_auth_trigger_template_1.generateNestedAuthTriggerTemplate(this._category, this.resourceName, this._cognitoStackProps);
        }
        await this.generateStackResources(this._cognitoStackProps);
        await this.applyOverride();
        const template = await this.synthesizeTemplates();
        await this.saveBuildFiles(context, template);
        return template;
    }
    async generateStackResources(props) {
        this.addCfnParameters(props);
        this.addCfnConditions(props);
        this._authTemplateObj.generateCognitoStackResources(props);
        this.generateCfnOutputs(props);
    }
}
exports.AmplifyAuthTransform = AmplifyAuthTransform;
