// @aws-amplify/amplify-category-auth/lib/provider-utils/awscloudformation/auth-stack-builder/auth-user-pool-group-stack-builder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmplifyUserPoolGroupStackOutputs = exports.getCfnParamslogicalId = exports.AmplifyUserPoolGroupStack = void 0;
const cdk = __importStar(require("@aws-cdk/core"));
const iam = __importStar(require("@aws-cdk/aws-iam"));
const lambda = __importStar(require("@aws-cdk/aws-lambda"));
const aws_cognito_1 = require("@aws-cdk/aws-cognito");
const fs = __importStar(require("fs-extra"));
const constants_1 = require("../constants");
const CFN_TEMPLATE_FORMAT_VERSION = '2010-09-09';
const ROOT_CFN_DESCRIPTION = 'Root Stack for AWS Amplify CLI';
class AmplifyUserPoolGroupStack extends cdk.Stack {
    constructor(scope, id, props) {
        super(scope, id, props);
        this._cfnParameterMap = new Map();
        this._cfnConditionMap = new Map();
        this.renderCloudFormationTemplate = (_) => {
            return JSON.stringify(this._toCloudFormation(), undefined, 2);
        };
        this.generateUserPoolGroupResources = async (props) => {
            props.groups.forEach(group => {
                this.userPoolGroup[`${group.groupName}`] = new aws_cognito_1.CfnUserPoolGroup(this, `${group.groupName}Group`, {
                    userPoolId: this.getCfnParameter(exports.getCfnParamslogicalId(props.cognitoResourceName, 'UserPoolId')).valueAsString,
                    groupName: group.groupName,
                    precedence: group.precedence,
                });
                this.userPoolGroup[`${group.groupName}`].description = 'override success';
                if (props.identityPoolName) {
                    this.userPoolGroup[`${group.groupName}`].addPropertyOverride('RoleArn', cdk.Fn.getAtt(`${group.groupName}GroupRole`, 'Arn').toString());
                    this.userPoolGroupRole[`${group.groupName}`] = new iam.CfnRole(this, `${group.groupName}GroupRole`, {
                        roleName: cdk.Fn.join('', [
                            this.getCfnParameter(exports.getCfnParamslogicalId(props.cognitoResourceName, 'UserPoolId')).valueAsString,
                            `-${group.groupName}GroupRole`,
                        ]),
                        assumeRolePolicyDocument: {
                            Version: '2012-10-17',
                            Statement: [
                                {
                                    Sid: '',
                                    Effect: 'Allow',
                                    Principal: {
                                        Federated: 'cognito-identity.amazonaws.com',
                                    },
                                    Action: 'sts:AssumeRoleWithWebIdentity',
                                    Condition: {
                                        StringEquals: {
                                            'cognito-identity.amazonaws.com:aud': {
                                                Ref: `auth${props.cognitoResourceName}IdentityPoolId`,
                                            },
                                        },
                                        'ForAnyValue:StringLike': { 'cognito-identity.amazonaws.com:amr': 'authenticated' },
                                    },
                                },
                            ],
                        },
                    });
                    if (group.customPolicies && group.customPolicies.length > 0) {
                        this.userPoolGroupRole[`${group.groupName}`].addPropertyOverride('Policies', JSON.stringify(group.customPolicies, null, 4));
                    }
                }
            });
            if (props.identityPoolName) {
                this.lambdaExecutionRole = new iam.CfnRole(this, 'LambdaExecutionRole', {
                    roleName: cdk.Fn.conditionIf('ShouldNotCreateEnvResources', props.cognitoResourceName, cdk.Fn.join('', [`${props.cognitoResourceName}-ExecutionRole-`, cdk.Fn.ref('env')]).toString()).toString(),
                    assumeRolePolicyDocument: {
                        Version: '2012-10-17',
                        Statement: [
                            {
                                Effect: 'Allow',
                                Principal: {
                                    Service: ['lambda.amazonaws.com'],
                                },
                                Action: ['sts:AssumeRole'],
                            },
                        ],
                    },
                    policies: [
                        {
                            policyName: 'UserGroupLogPolicy',
                            policyDocument: {
                                Version: '2012-10-17',
                                Statement: [
                                    {
                                        Effect: 'Allow',
                                        Action: ['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents'],
                                        Resource: 'arn:aws:logs:*:*:*',
                                    },
                                ],
                            },
                        },
                        {
                            policyName: 'UserGroupExecutionPolicy',
                            policyDocument: {
                                Version: '2012-10-17',
                                Statement: [
                                    {
                                        Effect: 'Allow',
                                        Action: [
                                            'cognito-identity:SetIdentityPoolRoles',
                                            'cognito-identity:ListIdentityPools',
                                            'cognito-identity:describeIdentityPool',
                                        ],
                                        Resource: '*',
                                    },
                                ],
                            },
                        },
                        {
                            policyName: 'UserGroupPassRolePolicy',
                            policyDocument: {
                                Version: '2012-10-17',
                                Statement: [
                                    {
                                        Effect: 'Allow',
                                        Action: ['iam:PassRole'],
                                        Resource: [
                                            {
                                                Ref: 'AuthRoleArn',
                                            },
                                            {
                                                Ref: 'UnauthRoleArn',
                                            },
                                        ],
                                    },
                                ],
                            },
                        },
                    ],
                });
                this.roleMapLambdaFunction = new lambda.CfnFunction(this, 'RoleMapFunction', {
                    code: {
                        zipFile: fs.readFileSync(constants_1.roleMapLambdaFilePath, 'utf-8'),
                    },
                    handler: 'index.handler',
                    runtime: 'nodejs12.x',
                    timeout: 300,
                    role: cdk.Fn.getAtt('LambdaExecutionRole', 'Arn').toString(),
                });
                this.roleMapCustomResource = new cdk.CustomResource(this, 'RoleMapFunctionInput', {
                    serviceToken: this.roleMapLambdaFunction.attrArn,
                    resourceType: 'Custom::LambdaCallout',
                    properties: {
                        AuthRoleArn: cdk.Fn.ref('AuthRoleArn'),
                        UnauthRoleArn: cdk.Fn.ref('UnauthRoleArn'),
                        identityPoolId: cdk.Fn.ref(exports.getCfnParamslogicalId(props.cognitoResourceName, 'IdentityPoolId')),
                        userPoolId: cdk.Fn.ref(exports.getCfnParamslogicalId(props.cognitoResourceName, 'UserPoolId')),
                        appClientIDWeb: cdk.Fn.ref(exports.getCfnParamslogicalId(props.cognitoResourceName, 'AppClientIDWeb')),
                        appClientID: cdk.Fn.ref(exports.getCfnParamslogicalId(props.cognitoResourceName, 'AppClientID')),
                        region: cdk.Fn.ref('AWS::Region'),
                        env: cdk.Fn.ref('env'),
                    },
                });
                this.roleMapCustomResource.node.addDependency(this.roleMapLambdaFunction);
            }
        };
        this._scope = scope;
        this.templateOptions.templateFormatVersion = CFN_TEMPLATE_FORMAT_VERSION;
        this.templateOptions.description = ROOT_CFN_DESCRIPTION;
        this.userPoolGroup = {};
        this.userPoolGroupRole = {};
    }
    getCfnOutput(logicalId) {
        throw new Error('Method not implemented.');
    }
    getCfnMapping(logicalId) {
        throw new Error('Method not implemented.');
    }
    addCfnOutput(props, logicalId) {
        try {
            new cdk.CfnOutput(this, logicalId, props);
        }
        catch (error) {
            throw new Error(error);
        }
    }
    addCfnMapping(props, logicalId) {
        try {
            new cdk.CfnMapping(this, logicalId, props);
        }
        catch (error) {
            throw new Error(error);
        }
    }
    addCfnResource(props, logicalId) {
        try {
            new cdk.CfnResource(this, logicalId, props);
        }
        catch (error) {
            throw new Error(error);
        }
    }
    addCfnParameter(props, logicalId) {
        try {
            if (this._cfnParameterMap.has(logicalId)) {
                throw new Error('logical Id already Exists');
            }
            this._cfnParameterMap.set(logicalId, new cdk.CfnParameter(this, logicalId, props));
        }
        catch (error) {
            throw new Error(error);
        }
    }
    addCfnCondition(props, logicalId) {
        try {
            if (this._cfnConditionMap.has(logicalId)) {
                throw new Error('logical Id already Exists');
            }
            this._cfnConditionMap.set(logicalId, new cdk.CfnCondition(this, logicalId, props));
        }
        catch (error) {
            throw new Error(error);
        }
    }
    getCfnParameter(logicalId) {
        if (this._cfnParameterMap.has(logicalId)) {
            return this._cfnParameterMap.get(logicalId);
        }
        else {
            throw new Error(`Cfn Parameter with LogicalId ${logicalId} doesnt exist`);
        }
    }
    getCfnCondition(logicalId) {
        if (this._cfnConditionMap.has(logicalId)) {
            return this._cfnConditionMap.get(logicalId);
        }
        else {
            throw new Error(`Cfn Parameter with LogicalId ${logicalId} doesnt exist`);
        }
    }
}
exports.AmplifyUserPoolGroupStack = AmplifyUserPoolGroupStack;
const getCfnParamslogicalId = (cognitoResourceName, cfnParamName) => {
    return `auth${cognitoResourceName}${cfnParamName}`;
};
exports.getCfnParamslogicalId = getCfnParamslogicalId;
class AmplifyUserPoolGroupStackOutputs extends cdk.Stack {
    constructor(scope, id, props) {
        super(scope, id, props);
        this.renderCloudFormationTemplate = (_) => {
            return JSON.stringify(this._toCloudFormation(), undefined, 2);
        };
    }
    getCfnParameter(logicalId) {
        throw new Error('Method not implemented.');
    }
    getCfnOutput(logicalId) {
        throw new Error('Method not implemented.');
    }
    getCfnMapping(logicalId) {
        throw new Error('Method not implemented.');
    }
    getCfnCondition(logicalId) {
        throw new Error('Method not implemented.');
    }
    addCfnParameter(props, logicalId) {
        throw new Error('Method not implemented.');
    }
    addCfnOutput(props, logicalId) {
        try {
            new cdk.CfnOutput(this, logicalId, props);
        }
        catch (error) {
            throw new Error(error);
        }
    }
    addCfnMapping(props, logicalId) {
        throw new Error('Method not implemented.');
    }
    addCfnCondition(props, logicalId) {
        throw new Error('Method not implemented.');
    }
    addCfnResource(props, logicalId) {
        throw new Error('Method not implemented.');
    }
}
exports.AmplifyUserPoolGroupStackOutputs = AmplifyUserPoolGroupStackOutputs;
