// @aws-amplify/amplify-category-auth/lib/provider-utils/awscloudformation/auth-stack-builder/user-pool-group-stack-transform.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmplifyUserPoolGroupTransform = void 0;
const amplify_cli_core_1 = require("amplify-cli-core");
const stack_synthesizer_1 = require("./stack-synthesizer");
const cdk = __importStar(require("@aws-cdk/core"));
const auth_input_state_1 = require("../auth-inputs-manager/auth-input-state");
const path = __importStar(require("path"));
const auth_user_pool_group_stack_builder_1 = require("./auth-user-pool-group-stack-builder");
const amplify_prompts_1 = require("amplify-prompts");
const fs = __importStar(require("fs-extra"));
const vm = __importStar(require("vm2"));
const os_1 = __importDefault(require("os"));
class AmplifyUserPoolGroupTransform extends amplify_cli_core_1.AmplifyCategoryTransform {
    constructor(resourceName) {
        super(resourceName);
        this.generateStackResources = async (props) => {
            this._userPoolGroupTemplateObj = new auth_user_pool_group_stack_builder_1.AmplifyUserPoolGroupStack(this._app, 'AmplifyUserPoolGroupStack', {
                synthesizer: this._synthesizer,
            });
            this.__userPoolGroupTemplateObjOutputs = new auth_user_pool_group_stack_builder_1.AmplifyUserPoolGroupStackOutputs(this._app, 'AmplifyUserPoolGroupStackOutputs', {
                synthesizer: this._synthesizerOutputs,
            });
            this._userPoolGroupTemplateObj.addCfnParameter({
                type: 'String',
            }, 'env');
            this._userPoolGroupTemplateObj.addCfnParameter({
                type: 'String',
            }, 'AuthRoleArn');
            this._userPoolGroupTemplateObj.addCfnParameter({
                type: 'String',
            }, 'UnauthRoleArn');
            this._userPoolGroupTemplateObj.addCfnParameter({
                type: 'String',
                default: `auth${props.cognitoResourceName}UserPoolId`,
            }, `auth${props.cognitoResourceName}UserPoolId`);
            if (props.identityPoolName) {
                this._userPoolGroupTemplateObj.addCfnParameter({
                    type: 'String',
                    default: `auth${props.cognitoResourceName}IdentityPoolId`,
                }, `auth${props.cognitoResourceName}IdentityPoolId`);
            }
            this._userPoolGroupTemplateObj.addCfnParameter({
                type: 'String',
                default: `auth${props.cognitoResourceName}AppClientID`,
            }, `auth${props.cognitoResourceName}AppClientID`);
            this._userPoolGroupTemplateObj.addCfnParameter({
                type: 'String',
                default: `auth${props.cognitoResourceName}AppClientIDWeb`,
            }, `auth${props.cognitoResourceName}AppClientIDWeb`);
            this._userPoolGroupTemplateObj.addCfnCondition({
                expression: cdk.Fn.conditionEquals(this._userPoolGroupTemplateObj.getCfnParameter('env'), 'NONE'),
            }, 'ShouldNotCreateEnvResources');
            await this._userPoolGroupTemplateObj.generateUserPoolGroupResources(props);
            if (props.identityPoolName) {
                props.groups.forEach(group => {
                    this.__userPoolGroupTemplateObjOutputs.addCfnOutput({
                        value: cdk.Fn.getAtt(`${group.groupName}GroupRole`, 'Arn').toString(),
                    }, `${group.groupName}GroupRole`);
                });
            }
        };
        this.applyOverride = async () => {
            const backendDir = amplify_cli_core_1.pathManager.getBackendDirPath();
            const overrideDir = path.join(backendDir, this._category, this._resourceName);
            const isBuild = await amplify_cli_core_1.buildOverrideDir(backendDir, overrideDir).catch(error => {
                amplify_prompts_1.printer.debug(`Skipping build as ${error.message}`);
                return false;
            });
            if (isBuild) {
                const overrideCode = await fs.readFile(path.join(overrideDir, 'build', 'override.js'), 'utf-8').catch(() => {
                    amplify_prompts_1.formatter.list(['No override File Found', `To override ${this._resourceName} run amplify override auth`]);
                    return '';
                });
                const sandboxNode = new vm.NodeVM({
                    console: 'inherit',
                    timeout: 5000,
                    sandbox: {},
                });
                try {
                    sandboxNode.run(overrideCode).override(this._userPoolGroupTemplateObj);
                }
                catch (err) {
                    const error = new Error(`Skipping override due to ${err}${os_1.default.EOL}`);
                    amplify_prompts_1.printer.error(`${error}`);
                    error.stack = undefined;
                    throw error;
                }
            }
        };
        this.generateStackProps = async (context) => {
            const resourceDirPath = path.join(amplify_cli_core_1.pathManager.getBackendDirPath(), 'auth', 'userPoolGroups', 'user-pool-group-precedence.json');
            const groups = amplify_cli_core_1.JSONUtilities.readJson(resourceDirPath, { throwIfNotExist: true });
            const cliState = new auth_input_state_1.AuthInputState(this._authResourceName);
            this._cliInputs = cliState.getCLIInputPayload();
            const identityPoolName = this._cliInputs.cognitoConfig.identityPoolName;
            return {
                groups: groups,
                identityPoolName,
                cognitoResourceName: this._authResourceName,
            };
        };
        this.synthesizeTemplates = async () => {
            this._app.synth();
            const templates = this._synthesizer.collectStacks();
            const cfnUserPoolGroupStack = templates.get('AmplifyUserPoolGroupStack');
            const templatesOutput = this._synthesizerOutputs.collectStacks();
            const cfnUserPoolGroupOutputs = templatesOutput.get('AmplifyUserPoolGroupStackOutputs');
            cfnUserPoolGroupStack.Outputs = cfnUserPoolGroupOutputs.Outputs;
            return cfnUserPoolGroupStack;
        };
        this.saveBuildFiles = async (context, template) => {
            const cognitoStackFileName = `${this._resourceName}-cloudformation-template.json`;
            const cognitostackFilePath = path.join(amplify_cli_core_1.pathManager.getBackendDirPath(), this._category, this._resourceName, 'build', cognitoStackFileName);
            await amplify_cli_core_1.writeCFNTemplate(template, cognitostackFilePath, {
                templateFormat: amplify_cli_core_1.CFNTemplateFormat.JSON,
            });
            this.writeBuildFiles(context);
        };
        this.writeBuildFiles = async (context) => {
            const parametersJSONFilePath = path.join(amplify_cli_core_1.pathManager.getBackendDirPath(), this._category, this._resourceName, 'build', 'parameters.json');
            const roles = {
                AuthRoleArn: {
                    'Fn::GetAtt': ['AuthRole', 'Arn'],
                },
                UnauthRoleArn: {
                    'Fn::GetAtt': ['UnauthRole', 'Arn'],
                },
            };
            let parameters = {
                ...roles,
            };
            amplify_cli_core_1.JSONUtilities.writeJson(parametersJSONFilePath, parameters);
        };
        this._authResourceName = resourceName;
        this._resourceName = 'userPoolGroups';
        this._synthesizer = new stack_synthesizer_1.AuthStackSythesizer();
        this._synthesizerOutputs = new stack_synthesizer_1.AuthStackSythesizer();
        this._app = new cdk.App();
        this._category = amplify_cli_core_1.AmplifyCategories.AUTH;
        this._service = amplify_cli_core_1.AmplifySupportedService.COGNITOUSERPOOLGROUPS;
    }
    async transform(context) {
        const userPoolGroupStackOptions = await this.generateStackProps(context);
        await this.generateStackResources(userPoolGroupStackOptions);
        await this.applyOverride();
        const template = await this.synthesizeTemplates();
        await this.saveBuildFiles(context, template);
        return template;
    }
}
exports.AmplifyUserPoolGroupTransform = AmplifyUserPoolGroupTransform;
